smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/core/CodeWriterExt.kt

```diff
@@ -0,0 +1,141 @@
+ * the provided function `openBlock(String textBeforeNewline, String textAfterNewline, Runnable r)`
+ *
+ * Example:
+ * ```
+ * writer.withBlock("{", "}") {
+ *     write("foo")
+ * }
+ * ```
+ *
+ * Equivalent to:
+ * ```
+ * writer.openBlock("{")
+ * writer.write("foo")
+ * writer.closeBlock("}")
+ * ```
+ */
+fun <T : CodeWriter> T.withBlock(
+    textBeforeNewLine: String,
+    textAfterNewLine: String,
+    vararg args: Any,
+    block: T.() -> Unit
+): T = wrapBlockIf(true, textBeforeNewLine, textAfterNewLine, *args) { block(this) }
+
+/**
+ * Extension function that is more idiomatic Kotlin that is roughly the same purpose as an if block wrapped around
+ * the provided function `openBlock(String textBeforeNewline, String textAfterNewline, Runnable r)`
+ *
+ * Example:
+ * ```
+ * writer.wrapBlockIf(foo == bar, "{", "}") {
+ *     write("foo")
+ * }
+ * ```
+ *
+ * Equivalent to:
+ * ```
+ * if (foo == bar) writer.openBlock("{")
+ * writer.write("foo")
+ * if (foo == bar) writer.closeBlock("}")
+ * ```
+ */
+fun <T : CodeWriter> T.wrapBlockIf(
+    condition: Boolean,
+    textBeforeNewLine: String,
+    textAfterNewLine: String,
+    vararg args: Any,
+    block: T.() -> Unit,
+): T {
+    if (condition) openBlock(textBeforeNewLine, *args)
+    block(this)
+    if (condition) closeBlock(textAfterNewLine)
+    return this
+}
+
+/**
+ * Extension function that closes the previous block, dedents, opens a new block with [textBeforeNewLine], and indents.
+ *
+ * This is useful for chaining if-if-else-else branches.
+ *
+ * Example:
+ * ```
+ * writer.openBlock("if (foo) {")
+ *     .write("foo()")
+ *     .closeAndOpenBlock("} else {")
+ *     .write("bar()")
+ *     .closeBlock("}")
+ * ```
+ */
+fun <T : CodeWriter> T.closeAndOpenBlock(
+    textBeforeNewLine: String,
+    vararg args: Any,
+): T = apply {
+    dedent()
+    openBlock(textBeforeNewLine, *args)
+}
+
+/**
+ * Declares a section for extension in codegen.  The [SectionId] should be specified as a child
+ * of the type housing the codegen associated with the section. This keeps [SectionId]s closely
+ * associated with their targets.
+ */
+fun <T : CodeWriter> T.declareSection(id: SectionId, context: Map<String, Any?> = emptyMap(), block: T.() -> Unit = {}): T {
+    putContext(context)
+    pushState(id.javaClass.canonicalName)
+    block(this)
+    popState()
+    removeContext(context)
+    return this
+}
+
+private fun <T : CodeWriter> T.removeContext(context: Map<String, Any?>): Unit =
+    context.keys.forEach { key -> removeContext(key) }
+
+/**
+ * Convenience function to get a typed value out of the context or throw if the key doesn't exist
+ * or the type is wrong
+ */
+inline fun <reified T> CodeWriter.getContextValue(key: String): T = checkNotNull(getContext(key) as? T) {
+    "Expected `$key` in CodeWriter context"
+}
+
+// Specifies a function that receives a [CodeWriter]
+typealias InlineCodeWriter = CodeWriter.() -> Unit
```

Comment:
nit

Still think `Writable` may help connect to the formatter `W` better as far as cognitive overhead

Reply:
This issue is that there are two, one for `CodeWriter` and one for `KotlinWriter`.  Do you prefer `KotlinWritable` and `CodeWritable` to what is here?

Reply:
Meh with two of them I guess I don't care as much. It's fine as is